calico/felix/endpoint.py

```diff
@@ -431,75 +426,63 @@ def __str__(self):
                  self._iface_name or "unknown"))
 
 
-def _get_endpoint_rules(endpoint_id, suffix, ip_version, local_ips, mac,
-                        profile_ids):
+def _get_endpoint_rules(endpoint_id, suffix, mac, profile_ids):
     to_chain_name, from_chain_name = chain_names(suffix)
 
-    # First build the chain that manages packets to the interface.
-    # Common chain prefixes.
-    to_chain = []
+    to_chain, to_deps = _build_to_or_from_chain(
+        endpoint_id,
+        profile_ids,
+        to_chain_name,
+        "inbound"
+    )
+    from_chain, from_deps = _build_to_or_from_chain(
+        endpoint_id,
+        profile_ids,
+        from_chain_name,
+        "outbound",
+        expected_mac=mac,
+    )
+
+    updates = {to_chain_name: to_chain, from_chain_name: from_chain}
+    deps = {to_chain_name: to_deps, from_chain_name: from_deps}
+    return updates, deps
+
 
+def _build_to_or_from_chain(endpoint_id, profile_ids, chain_name,
+                            direction, expected_mac=None):
+    # Ensure the MARK is set to 0 when we start so that unmatched packets will
+    # be dropped.
+    chain = [
+        "--append %s --jump MARK --set-mark 0" % chain_name
+    ]
+    if expected_mac:
+        _log.debug("Policing source MAC: %s", expected_mac)
+        chain.append('--append %s --match mac ! --mac-source %s --jump DROP '
+                     '--match comment --comment "Incorrect source MAC"' %
+                     (chain_name, expected_mac))
     # Jump to each profile in turn.  The profile will do one of the
     # following:
     # * DROP the packet; in which case we won't see it again.
-    # * RETURN without MARKing the packet; indicates that the packet was
-    #   accepted.
-    # * RETURN with a mark on the packet, indicates reaching the end
-    #   of the chain.
-    to_deps = set()
+    # * RETURN the packet with MARK==1, indicating it accepted the packet. In
+    #   which case, we RETURN and skip further profiles.
+    # * RETURN the packet with MARK==0, indicating it did not match the packet.
+    #   In which case, we carry on and process the next profile.
+    deps = set()
     for profile_id in profile_ids:
-        profile_in_chain = profile_to_chain_name("inbound", profile_id)
-        to_deps.add(profile_in_chain)
-        to_chain.append("--append %s --jump MARK --set-mark 0" %
-                        to_chain_name)
-        to_chain.append("--append %s --jump %s" %
-                        (to_chain_name, profile_in_chain))
-        to_chain.append('--append %s --match mark ! --mark 1/1 '
-                        '--match comment --comment "No mark means profile '
-                        'accepted packet" --jump RETURN' % to_chain_name)
+        profile_in_chain = profile_to_chain_name(direction, profile_id)
```

Comment:
Maybe rename profile_in_chain to profile_chain, given that it's not now necessarily inbound?